File: sdks/go/autonoma/unique.go
package autonoma

import (
	"crypto/sha256"
	"encoding/hex"
	"regexp"
	"strings"
)

// Deterministic uniqueness helpers seeded from testRunId. A scenario's data
// needs stable keys across runs but unique values per run (unique emails, org
// slugs, ids). These derive that uniqueness from (testRunId, ...parts): the
// same inputs always produce the same output within a run, so a scenario's Up
// and a later Down compute identical values without storing them.
//
// The recipe is sha256(testRunId + (" " + part) for each part), hex-encoded,
// truncated to the first 12 chars - and MUST match the other language SDKs
// byte-for-byte for cross-language conformance.

const uniqueTokenLength = 12

var (
	slugNonAlnum    = regexp.MustCompile(`[^a-z0-9]+`)
	slugTrimHyphens = regexp.MustCompile(`^-+|-+$`)
)

func uniqueDigest(testRunID string, parts ...string) string {
	h := sha256.New()
	h.Write([]byte(testRunID))
	for _, part := range parts {
		h.Write([]byte(" "))
		h.Write([]byte(part))
	}
	return hex.EncodeToString(h.Sum(nil))
}

// UniqueToken returns a short hex token, deterministic per (testRunId, ...parts).
func UniqueToken(testRunID string, parts ...string) string {
	return uniqueDigest(testRunID, parts...)[:uniqueTokenLength]
}

// UniqueID returns an id like "user_1a2b3c4d5e6f", deterministic per inputs.
// An empty prefix defaults to "id".
func UniqueID(testRunID string, prefix string, parts ...string) string {
	if prefix == "" {
		prefix = "id"
	}
	tokenParts := append([]string{prefix}, parts...)
	return prefix + "_" + UniqueToken(testRunID, tokenParts...)
}

// UniqueSlug returns a URL-safe slug like "acme-1a2b3c4d5e6f", deterministic
// per inputs. An empty base defaults to "item".
func UniqueSlug(testRunID string, base string, parts ...string) string {
	if base == "" {
		base = "item"
	}
	tokenParts := append([]string{base}, parts...)
	token := UniqueToken(testRunID, tokenParts...)
	normalized := slugTrimHyphens.ReplaceAllString(
		slugNonAlnum.ReplaceAllString(strings.ToLower(base), "-"), "")
	if normalized == "" {
		normalized = "item"
	}
	return normalized + "-" + token
}

// UniqueEmail returns an email like "user+1a2b3c4d5e6f@example.com",
// deterministic per inputs. Empty local/domain default to "user"/"example.com".
func UniqueEmail(testRunID string, local string, domain string) string {
	if local == "" {
		local = "user"
	}
	if domain == "" {
		domain = "example.com"
	}
	return local + "+" + UniqueToken(testRunID, local, domain) + "@" + domain
}

File: sdks/python/src/autonoma/unique.py
"""Deterministic uniqueness helpers seeded from ``test_run_id``.

A scenario's ``data`` must have stable *keys* across runs, but its *values*
should be unique per run (unique emails, org slugs, ids) so two concurrent
runs never collide. These helpers derive that uniqueness deterministically
from ``(test_run_id, *inputs)``: the same inputs always produce the same
output within a run, so a scenario's ``up`` and a later ``down`` compute
identical values without storing them.
"""

from __future__ import annotations

import hashlib
import re

_TOKEN_LENGTH = 12
_SLUG_STRIP = re.compile(r"[^a-z0-9]+")


def _digest(test_run_id: str, parts: tuple[object, ...]) -> str:
    h = hashlib.sha256()
    h.update(test_run_id.encode("utf-8"))
    for part in parts:
        h.update(b" ")
        h.update(str(part).encode("utf-8"))
    return h.hexdigest()


def unique_token(test_run_id: str, *parts: object) -> str:
    """A short hex token, deterministic per ``(test_run_id, *parts)``."""
    return _digest(test_run_id, parts)[:_TOKEN_LENGTH]


def unique_id(test_run_id: str, prefix: str = "id", *parts: object) -> str:
    """A unique id like ``user_1a2b3c4d5e6f``, deterministic per inputs."""
    return f"{prefix}_{unique_token(test_run_id, prefix, *parts)}"


def unique_slug(test_run_id: str, base: str = "item", *parts: object) -> str:
    """A URL-safe slug like ``acme-1a2b3c4d5e6f``, deterministic per inputs."""
    normalized = _SLUG_STRIP.sub("-", str(base).lower()).strip("-") or "item"
    return f"{normalized}-{unique_token(test_run_id, base, *parts)}"


def unique_email(
    test_run_id: str, local: str = "user", domain: str = "example.com"
) -> str:
    """A unique email like ``user+1a2b3c4d5e6f@example.com``, deterministic per inputs."""
    return f"{local}+{unique_token(test_run_id, local, domain)}@{domain}"
